### PR TITLE
Add integration tests for Option

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/rust-diplomat/diplomat/issues/92

I think this should be a decent set of tests for now. There's more stuff that could be tested but that functionality does not exist yet.

### CODE REVIEW

File: core/src/ast/validity.rs [hunk 1]
```diff
@@ -124,6 +129,50 @@ mod tests {
...
+
+                    }
+                }
+            }
+        };
+    }
+
+    #[test]
+    fn option_valid() {
+        uitest_validity! {
```
Comment (sffc) [inline core/src/ast/validity.rs:160]: Nit: Add a test for an option of a reference

File: tool/src/js/conversions.rs [hunk 2]
```diff
@@ -54,6 +54,10 @@ pub fn gen_value_js_to_rust(
                 param_name, param_name, align
             ));
         }
+        ast::TypeName::Primitive(PrimitiveType::char) => {
+            pre_logic.push(format!("if (!{p}.length || !{p}.codePointAt || {p}.length != 1) {{ throw new TypeError(\"Expected single-character string for `char` parameter {p}, found \" + {p}); }}", p=param_name));
```
Comment (sffc) [inline tool/src/js/conversions.rs:58]: Issue: `"😋".length === 2`

File: feature_tests/js/api.mjs [hunk 3]
```diff
@@ -105,7 +105,187 @@ export class Opaque {
...
+  constructor(underlying) {
+    this.underlying = underlying;
+  }
+
+  static new(i) {
+    const diplomat_out = (() => {
+      if (wasm.OptionOpaque_new(i) !== 0) {
+        const inhabited_value = (() => {
+          const out = (() => {
+            const out = new OptionOpaque(wasm.OptionOpaque_new(i));
```
Comment (sffc) [inline feature_tests/js/api.mjs:127]: Question: why are you calling `wasm.OptionOpaque_new(i)` twice?  Shouldn't you call it once, on line 124, and save the value for down here?

File: tool/src/js/conversions.rs (hunk 2 above)
Comment (Manishearth) [reply tool/src/js/conversions.rs]: Fortunately JS string _iteration_ does the sensible thing here

File: feature_tests/js/api.mjs [hunk 4]
```diff
@@ -105,7 +105,7 @@ export class Opaque {
     const diplomat_MyStruct_extracted_d = s["d"];
     const diplomat_MyStruct_extracted_e = s["e"];
     const diplomat_MyStruct_extracted_f = s["f"];
-    if (!diplomat_MyStruct_extracted_f.length || !diplomat_MyStruct_extracted_f.codePointAt || diplomat_MyStruct_extracted_f.length != 1) { throw new TypeError("Expected single-character string for `char` parameter diplomat_MyStruct_extracted_f, found " + diplomat_MyStruct_extracted_f); }
+    if (!diplomat_MyStruct_extracted_f.length || !diplomat_MyStruct_extracted_f.codePointAt || [...diplomat_MyStruct_extracted_f].length != 1) { throw new TypeError("Expected single-character string for `char` parameter diplomat_MyStruct_extracted_f, found " + diplomat_MyStruct_extracted_f); }
     const diplomat_out = wasm.Opaque_assert_struct(this.underlying, diplomat_MyStruct_extracted_a, diplomat_MyStruct_extracted_b, diplomat_MyStruct_extracted_c, diplomat_MyStruct_extracted_d, diplomat_MyStruct_extracted_e, diplomat_MyStruct_extracted_f.codePointAt(0));
```
Comment (sffc) [inline feature_tests/js/api.mjs:109]: Nit (optional): This is JavaScript.  I'm not a fan of this brand checking.  How about using optional chaining and then checking for undefined, like

```javascript
const cp = diplomat_MyStruct_extracted_f?.codePointAt(0);
if (!cp && cp !== 0) {
    // throw TypeError
}
// use cp
```

**Review by sffc — APPROVED** [review]

**Review by sffc — COMMENTED** [review]
Praise: Thanks for moving this into a helper function

**Pull request merged**